File: sdk/typescript/src/api/escrow.ts
import type { HttpClient } from "../http.js";
import type { TinyVerseWebSocket } from "../websocket.js";
import type {
  Escrow,
  EscrowCreateRequest,
  EscrowDispute,
  EscrowEvidence,
  EscrowMilestone,
  EscrowQueryParams,
} from "../types/index.js";

export class EscrowApi {
  constructor(
    private readonly http: HttpClient,
    private readonly wsFactory?: (
      path: string,
      options?: { directoryAuth?: boolean },
    ) => TinyVerseWebSocket,
  ) {}

  list(params?: EscrowQueryParams): Promise<{ escrows: Array<Escrow> }> {
    return this.http.getAuth<{ escrows: Array<Escrow> }>(
      "/escrow",
      params as Record<string, unknown>,
    );
  }

  create(request: EscrowCreateRequest): Promise<Escrow> {
    return this.http.postDirectoryAuthAs<Escrow>(
      "/escrow",
      request.client,
      request,
    );
  }

  get(escrowId: string): Promise<Escrow> {
    return this.http.getAuth<Escrow>(`/escrow/${encodeURIComponent(escrowId)}`);
  }

  stream(escrowId: string, agentId?: string): TinyVerseWebSocket | undefined {
    const query = agentId
      ? `?${new URLSearchParams({ "X-Agent-ID": agentId }).toString()}`
      : "";
    return this.wsFactory?.(
      `/escrow/${encodeURIComponent(escrowId)}/stream${query}`,
      agentId ? { directoryAuth: true } : undefined,
    );
  }

  accept(escrowId: string, actor?: string): Promise<Escrow> {
    return this.postEscrowActor(
      `/escrow/${encodeURIComponent(escrowId)}/accept`,
      actor,
    );
  }

  deliver(
    escrowId: string,
    proof: { actor?: string; description: string; refs?: Array<string> },
  ): Promise<Escrow> {
    return this.postEscrowActor(
      `/escrow/${encodeURIComponent(escrowId)}/deliver`,
      proof.actor,
      proof,
    );
  }

  acceptDelivery(
    escrowId: string,
    actor?: string,
    onChainTx?: string,
  ): Promise<Escrow> {
    return this.postEscrowActor(
      `/escrow/${encodeURIComponent(escrowId)}/accept-delivery`,
      actor,
      onChainTx ? { onChainTx } : undefined,
    );
  }

  claimRelease(
    escrowId: string,
    actor?: string,
    onChainTx?: string,
  ): Promise<Escrow> {
    return this.postEscrowActor(
      `/escrow/${encodeURIComponent(escrowId)}/claim-release`,
      actor,
      onChainTx ? { onChainTx } : undefined,
    );
  }

  claimRefund(
    escrowId: string,
    actor?: string,
    onChainTx?: string,
  ): Promise<Escrow> {
    return this.postEscrowActor(
      `/escrow/${encodeURIComponent(escrowId)}/claim-refund`,
      actor,
      onChainTx ? { onChainTx } : undefined,
    );
  }

  requestRevision(
    escrowId: string,
    reason: string,
    actor?: string,
  ): Promise<Escrow> {
    return this.postEscrowActor(
      `/escrow/${encodeURIComponent(escrowId)}/request-revision`,
      actor,
      { actor, reason },
    );
  }

  cancel(escrowId: string, actor?: string, onChainTx?: string): Promise<Escrow> {
    return this.postEscrowActor(
      `/escrow/${encodeURIComponent(escrowId)}/cancel`,
      actor,
      onChainTx ? { onChainTx } : undefined,
    );
  }

  extendDeadline(
    escrowId: string,
    newDeadline: string,
    actor?: string,
  ): Promise<Escrow> {
    return this.postEscrowActor(
      `/escrow/${encodeURIComponent(escrowId)}/extend-deadline`,
      actor,
      { actor, deadline: newDeadline },
    );
  }

  approveExtension(escrowId: string, actor?: string): Promise<Escrow> {
    return this.postEscrowActor(
      `/escrow/${encodeURIComponent(escrowId)}/approve-extension`,
      actor,
    );
  }

  // --- Disputes ---

  openDispute(
    escrowId: string,
    reason: string,
    actor?: string,
  ): Promise<EscrowDispute> {
    return this.postEscrowActor(
      `/escrow/${encodeURIComponent(escrowId)}/dispute`,
      actor,
      { actor, reason },
    );
  }

  getDispute(escrowId: string): Promise<EscrowDispute> {
    return this.http.getAuth<EscrowDispute>(`/escrow/${encodeURIComponent(escrowId)}/dispute`);
  }

  submitEvidence(
    escrowId: string,
    evidence: { actor?: string; type: string; description: string; ref?: string },
  ): Promise<void> {
    return this.postEscrowActor(
      `/escrow/${encodeURIComponent(escrowId)}/dispute/evidence`,
      evidence.actor,
      evidence,
    );
  }

  acceptMediation(escrowId: string, actor?: string): Promise<EscrowDispute> {
    return this.postEscrowActor(
      `/escrow/${encodeURIComponent(escrowId)}/dispute/accept-mediation`,
      actor,
    );
  }

  rejectMediation(escrowId: string, actor?: string): Promise<EscrowDispute> {
    return this.postEscrowActor(
      `/escrow/${encodeURIComponent(escrowId)}/dispute/reject-mediation`,
      actor,
    );
  }

  payArbitration(
    escrowId: string,
    onChainTx: string,
    actor?: string,
  ): Promise<EscrowDispute> {
    return this.postEscrowActor(
      `/escrow/${encodeURIComponent(escrowId)}/dispute/pay-arbitration`,
      actor,
      { actor, onChainTx },
    );
  }

  voteArbitration(
    escrowId: string,
    vote: { actor?: string; councilMember: string; vote: string; clientPct?: number; providerPct?: number; rationale?: string },
  ): Promise<void> {
    return this.postEscrowActor(
      `/escrow/${encodeURIComponent(escrowId)}/dispute/vote`,
      vote.actor ?? vote.councilMember,
      vote,
    );
  }

  // --- Milestones ---

  deliverMilestone(
    escrowId: string,
    milestoneId: string,
    proof: { actor?: string; description: string; refs?: Array<string> },
  ): Promise<EscrowMilestone> {
    return this.postEscrowActor(
      `/escrow/${encodeURIComponent(escrowId)}/milestones/${encodeURIComponent(milestoneId)}/deliver`,
      proof.actor,
      proof,
    );
  }

  acceptMilestoneDelivery(
    escrowId: string,
    milestoneId: string,
    actor?: string,
    onChainTx?: string,
  ): Promise<EscrowMilestone> {
    return this.postEscrowActor(
      `/escrow/${encodeURIComponent(escrowId)}/milestones/${encodeURIComponent(milestoneId)}/accept-delivery`,
      actor,
      onChainTx ? { onChainTx } : undefined,
    );
  }

  requestMilestoneRevision(
    escrowId: string,
    milestoneId: string,
    reason: string,
    actor?: string,
  ): Promise<EscrowMilestone> {
    return this.postEscrowActor(
      `/escrow/${encodeURIComponent(escrowId)}/milestones/${encodeURIComponent(milestoneId)}/request-revision`,
      actor,
      { actor, reason },
    );
  }

  disputeMilestone(
    escrowId: string,
    milestoneId: string,
    reason: string,
    actor?: string,
  ): Promise<EscrowDispute> {
    return this.postEscrowActor(
      `/escrow/${encodeURIComponent(escrowId)}/milestones/${encodeURIComponent(milestoneId)}/dispute`,
      actor,
      { actor, reason },
    );
  }

  private postEscrowActor<T>(
    path: string,
    actor?: string,
    body?: Record<string, unknown>,
  ): Promise<T> {
    if (actor) {
      return this.http.postDirectoryAuthAs<T>(path, actor, {
        ...(body ?? {}),
        actor,
      });
    }
    return this.http.post<T>(path, body);
  }
}

File: sdk/rust/src/api/escrow.rs
use std::sync::Arc;

use serde::{Deserialize, Serialize};

use crate::error::Result;
use crate::http::{HttpClient, to_body, to_query};
use crate::types::{Escrow, EscrowCreateRequest, EscrowDispute, EscrowMilestone, EscrowQueryParams};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EscrowListResponse {
    pub escrows: Vec<Escrow>,
}

pub struct EscrowApi {
    http: Arc<HttpClient>,
}

impl EscrowApi {
    pub fn new(http: Arc<HttpClient>) -> Self {
        Self { http }
    }

    pub async fn list(&self, params: Option<&EscrowQueryParams>) -> Result<EscrowListResponse> {
        let query = params.map(to_query).transpose()?;
        self.http.get_auth("/escrow", query.as_ref()).await
    }

    pub async fn create(&self, request: &EscrowCreateRequest) -> Result<Escrow> {
        self.http.post("/escrow", Some(&to_body(request)?)).await
    }

    pub async fn get(&self, escrow_id: &str) -> Result<Escrow> {
        let path = format!("/escrow/{}", urlencoding::encode(escrow_id));
        self.http.get_auth(&path, None).await
    }

    pub async fn accept(&self, escrow_id: &str) -> Result<Escrow> {
        let path = format!("/escrow/{}/accept", urlencoding::encode(escrow_id));
        self.http.post(&path, None).await
    }

    pub async fn deliver(
        &self,
        escrow_id: &str,
        description: &str,
        refs: Option<&[String]>,
    ) -> Result<Escrow> {
        let path = format!("/escrow/{}/deliver", urlencoding::encode(escrow_id));
        let mut body = serde_json::json!({ "description": description });
        if let Some(r) = refs {
            body["refs"] = serde_json::json!(r);
        }
        self.http.post(&path, Some(&body)).await
    }

    pub async fn accept_delivery(&self, escrow_id: &str) -> Result<Escrow> {
        let path = format!("/escrow/{}/accept-delivery", urlencoding::encode(escrow_id));
        self.http.post(&path, None).await
    }

    pub async fn claim_release(&self, escrow_id: &str) -> Result<Escrow> {
        let path = format!("/escrow/{}/claim-release", urlencoding::encode(escrow_id));
        self.http.post(&path, None).await
    }

    pub async fn claim_refund(&self, escrow_id: &str) -> Result<Escrow> {
        let path = format!("/escrow/{}/claim-refund", urlencoding::encode(escrow_id));
        self.http.post(&path, None).await
    }

    pub async fn request_revision(&self, escrow_id: &str, reason: &str) -> Result<Escrow> {
        let path = format!(
            "/escrow/{}/request-revision",
            urlencoding::encode(escrow_id)
        );
        let body = serde_json::json!({ "reason": reason });
        self.http.post(&path, Some(&body)).await
    }

    pub async fn cancel(&self, escrow_id: &str) -> Result<Escrow> {
        let path = format!("/escrow/{}/cancel", urlencoding::encode(escrow_id));
        self.http.post(&path, None).await
    }

    pub async fn extend_deadline(
        &self,
        escrow_id: &str,
        new_deadline: &str,
    ) -> Result<Escrow> {
        let path = format!(
            "/escrow/{}/extend-deadline",
            urlencoding::encode(escrow_id)
        );
        let body = serde_json::json!({ "newDeadline": new_deadline });
        self.http.post(&path, Some(&body)).await
    }

    pub async fn approve_extension(&self, escrow_id: &str) -> Result<Escrow> {
        let path = format!(
            "/escrow/{}/approve-extension",
            urlencoding::encode(escrow_id)
        );
        self.http.post(&path, None).await
    }

    pub async fn open_dispute(&self, escrow_id: &str, reason: &str) -> Result<EscrowDispute> {
        let path = format!("/escrow/{}/dispute", urlencoding::encode(escrow_id));
        let body = serde_json::json!({ "reason": reason });
        self.http.post(&path, Some(&body)).await
    }

    pub async fn get_dispute(&self, escrow_id: &str) -> Result<EscrowDispute> {
        let path = format!("/escrow/{}/dispute", urlencoding::encode(escrow_id));
        self.http.get_auth(&path, None).await
    }

    pub async fn submit_evidence(
        &self,
        escrow_id: &str,
        evidence_type: &str,
        description: &str,
        reference: Option<&str>,
    ) -> Result<()> {
        let path = format!(
            "/escrow/{}/dispute/evidence",
            urlencoding::encode(escrow_id)
        );
        let mut body = serde_json::json!({
            "type": evidence_type,
            "description": description
        });
        if let Some(r) = reference {
            body["ref"] = serde_json::json!(r);
        }
        self.http.post(&path, Some(&body)).await
    }

    pub async fn accept_mediation(&self, escrow_id: &str) -> Result<EscrowDispute> {
        let path = format!(
            "/escrow/{}/dispute/accept-mediation",
            urlencoding::encode(escrow_id)
        );
        self.http.post(&path, None).await
    }

    pub async fn reject_mediation(&self, escrow_id: &str) -> Result<EscrowDispute> {
        let path = format!(
            "/escrow/{}/dispute/reject-mediation",
            urlencoding::encode(escrow_id)
        );
        self.http.post(&path, None).await
    }

    pub async fn pay_arbitration(
        &self,
        escrow_id: &str,
        amount: &str,
    ) -> Result<EscrowDispute> {
        let path = format!(
            "/escrow/{}/dispute/pay-arbitration",
            urlencoding::encode(escrow_id)
        );
        let body = serde_json::json!({ "amount": amount });
        self.http.post(&path, Some(&body)).await
    }

    pub async fn vote_arbitration(
        &self,
        escrow_id: &str,
        vote: &serde_json::Value,
    ) -> Result<()> {
        let path = format!(
            "/escrow/{}/dispute/vote",
            urlencoding::encode(escrow_id)
        );
        self.http.post(&path, Some(vote)).await
    }

    pub async fn deliver_milestone(
        &self,
        escrow_id: &str,
        milestone_id: &str,
        description: &str,
        refs: Option<&[String]>,
    ) -> Result<EscrowMilestone> {
        let path = format!(
            "/escrow/{}/milestones/{}/deliver",
            urlencoding::encode(escrow_id),
            urlencoding::encode(milestone_id)
        );
        let mut body = serde_json::json!({ "description": description });
        if let Some(r) = refs {
            body["refs"] = serde_json::json!(r);
        }
        self.http.post(&path, Some(&body)).await
    }

    pub async fn accept_milestone_delivery(
        &self,
        escrow_id: &str,
        milestone_id: &str,
    ) -> Result<EscrowMilestone> {
        let path = format!(
            "/escrow/{}/milestones/{}/accept-delivery",
            urlencoding::encode(escrow_id),
            urlencoding::encode(milestone_id)
        );
        self.http.post(&path, None).await
    }

    pub async fn request_milestone_revision(
        &self,
        escrow_id: &str,
        milestone_id: &str,
        reason: &str,
    ) -> Result<EscrowMilestone> {
        let path = format!(
            "/escrow/{}/milestones/{}/request-revision",
            urlencoding::encode(escrow_id),
            urlencoding::encode(milestone_id)
        );
        let body = serde_json::json!({ "reason": reason });
        self.http.post(&path, Some(&body)).await
    }

    pub async fn dispute_milestone(
        &self,
        escrow_id: &str,
        milestone_id: &str,
        reason: &str,
    ) -> Result<EscrowDispute> {
        let path = format!(
            "/escrow/{}/milestones/{}/dispute",
            urlencoding::encode(escrow_id),
            urlencoding::encode(milestone_id)
        );
        let body = serde_json::json!({ "reason": reason });
        self.http.post(&path, Some(&body)).await
    }
}
